### ISSUE TITLE
The evolution of constexpr: compile-time lookup tables in C

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10333]
The evolution of constexpr: compile-time lookup tables in C++<br>
<br>
https://ift.tt/1i8c7nK<br>
<br>
<br>
<br>
<div><div>
<p>Lookup tables (LUTs) are an important resource for systems programming. They are the embodiment of the time-space tradeoff: precomputing results allow faster computation, up to O(1)!</p>
<p>My background: I’m a modern C++ enthusiast and a embedded systems developer. Recently, I had to implement two CRC tables, with different parameters. Instead of hard coding the tables, I went for my favorite kind of approach: computed at compile time. As the codebase was in C++11, it reminded me how far we’ve come.</p>
<p>So, in this article, we’ll review how to implement compile-time lookup tables in C++, for every Standard C++ published so far: C++98, C++03, C++11, C++14, C++17 and C++20. We’ll focus on the evolution of <code>constexpr</code>, but we’ll also see important additions to the language, such as <code>auto</code> and library functionality.</p>
<p><strong>Note</strong>: for this article, we’ll only consider features available on each Standard, even if some of the library features were possible to implement in previous versions of C++.</p>

<h2>
<a name="c-98-03-the-dark-ages" href="#c-98-03-the-dark-ages" title='Link to "C++98/03: The Dark Ages"'>🔗</a>C++98/03: The Dark Ages</h2>
<p>Before modern C++, we did not have ways of compute constant expressions with the own language.</p>
<p>All we had for this task was exactly what C has:</p>
<a></a>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br></pre></td>
<td><pre><span></span><br><span><span>extern</span> <span>const</span> <span>int</span> lut[<span>16</span>];</span><br><span></span><br><span></span><br><span><span>#<span>include</span> <span>"my_table.h"</span></span></span><br><span><span>const</span> <span>int</span> lut[<span>16</span>] = {<span>1</span>,<span>2</span>,<span>3</span>, , <span>16</span>};</span><br></pre></td>
</tr></tbody></table></figure>
<p>We declare a lookup table in our header. Due to how C++ linkage works, we need to use <code>extern</code> linkage and define the lookup table in a single translation unit (<code>my_table.cpp</code>).
To use the table, all we need is to include the <code>my_table.hpp</code> file and invoke <code>lut[index]</code>.</p>
<p>Alternatively, we can introduce a function for lookup and make the LUT an implementation detail:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br><span>8</span><br><span>9</span><br><span>10</span><br><span>11</span><br></pre></td>
<td><pre><span></span><br><span><span><span>int</span> <span>lookup</span><span>(<span>size_t</span> index)</span></span>;</span><br><span></span><br><span></span><br><span><span>#<span>include</span> <span>"my_table.h"</span></span></span><br><span><span>static</span> <span>const</span> <span>int</span> lut[<span>16</span>] = {<span>1</span>,<span>2</span>,<span>3</span>, , <span>16</span>};</span><br><span></span><br><span><span><span>int</span> <span>lookup</span><span>(<span>size_t</span> index)</span></span>{</span><br><span> assert(index &lt; <span>16</span>); </span><br><span> <span>return</span> lut[index];</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<h3>
<a name="how-do-we-compute-our-lookup-tables-in-c-98" href="#how-do-we-compute-our-lookup-tables-in-c-98" title='Link to "How do we compute our lookup tables in C++98?"'>🔗</a>How do we compute our lookup tables in C++98?</h3>
<p>One could adventure with the preprocessor, but the most common way is generating the <code>my_table.cpp</code> file with a scripting language.</p>
<p>We can definitely do better than that!</p>

<h2>
<a name="c-11-dawn-of-constexpr" href="#c-11-dawn-of-constexpr" title='Link to "C++11: Dawn of constexpr"'>🔗</a>C++11: Dawn of <code>constexpr</code>
</h2>
<p>C++11 was such a huge improvement over the previous versions, it changed many paradigms in the language. <strong>Modern C++</strong> was born.</p>
<h3>
<a name="a-limited-constexpr" href="#a-limited-constexpr" title='Link to "A limited constexpr"'>🔗</a>A limited <code>constexpr</code>
</h3>
<p>The focus of our article is this beautiful inclusion to the language, so let’s start with it!</p>
<p>Though its name isn’t the most elegant-sounding, the keyword perfectly describes what it’s supposed to do: allow definition of <em>constant expressions</em>.</p>
<p>Now we can create type-safe constants, replacing macros:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br></pre></td>
<td><pre><span><span>constexpr</span> <span>size_t</span> lut_size = <span>99</span>;</span><br><span><span>const</span> <span>int</span> lut[lut_size] = {...};</span><br></pre></td>
</tr></tbody></table></figure>
<p>We can also create functions that can be evaluated at compile time:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br></pre></td>
<td><pre><span><span><span>constexpr</span> <span>size_t</span> <span>lut_size</span><span>()</span></span>{</span><br><span> <span>return</span> <span>99</span>;</span><br><span>}</span><br><span></span><br><span><span>const</span> <span>int</span> lut[lut_size()] = {...};</span><br></pre></td>
</tr></tbody></table></figure>
<p>With functions generating constant expressions, we have now a better way of computing anything at compile time. <code>constexpr</code> is Turing-complete!</p>
<p>The main limitation with C++11’s <code>constexpr</code>, however, is that <strong>it only allows one return statement</strong>. No <code>if</code>‘s and <code>else</code>‘s, no variables. Only one return. So, we had a lot of recursive functions and ternary operators:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br></pre></td>
<td><pre><span><span><span>constexpr</span> <span>unsigned</span> <span>factorial</span><span>(<span>unsigned</span> n)</span></span>{</span><br><span> <span>return</span> (n &lt; <span>2</span>) ? <span>1</span> : (n * factorial(n<span>-1</span>));</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<h3>
<a name="auto" href="#auto" title='Link to "auto"'>🔗</a><code>auto</code>
</h3>
<p>C++ is a statically-type language. The compiler already knows the type of an expression. So why should we?</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br></pre></td>
<td><pre><span><span>template</span>&lt;<span>typename</span> T&gt;</span><br><span><span><span>void</span> <span>do_something</span><span>(<span>const</span> T&amp; t)</span></span>{</span><br><span> </span><br><span> <span>auto</span> tmp = compute(t);</span><br><span> do_something_else(tmp);</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p>We can also have <code>auto</code> as our return type. The problem? It’s not deduced, and <strong>we still need a trailing return type</strong>:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br></pre></td>
<td><pre><span>auto do_something() -&gt; int {</span><br><span> <span>return</span> <span>0</span>;</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<h3>
<a name="decltype-and-std-declval" href="#decltype-and-std-declval" title='Link to "decltype and std::declval"'>🔗</a><code>decltype</code> and <code>std::declval</code>
</h3>
<p>How do we deduce a return type of a generic function, when we cannot just declare it as <code>auto</code>? We can use <code>decltype</code> so the compiler can tell us:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br></pre></td>
<td><pre><span><span>template</span>&lt;<span>typename</span> T1, <span>typename</span> T2&gt;</span><br><span>auto add(const T1&amp; a, const T1&amp; b) -&gt; decltype(a + b) {</span><br><span> <span>return</span> a + b;</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p>If we need a variable in <code>decltype</code> that we don’t have declared, we can use <a rel="noopener" href="https://en.cppreference.com/w/cpp/utility/declval" target="_blank">the <code>std::declval</code> utility</a>, to pretend we have one:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br></pre></td>
<td><pre><span><span>using</span> other_type = ;</span><br><span></span><br><span><span>template</span>&lt;<span>typename</span> F&gt;</span><br><span>auto call(F&amp;&amp; f) -&gt; decltype(f(std::declval&lt;other_type&gt;())){</span><br><span> other_type t { };</span><br><span> <span>return</span> f(t);</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<h3>
<a name="std-array" href="#std-array" title='Link to "std::array"'>🔗</a><code>std::array</code>
</h3>
<p>One issue with C-style arrays is that we cannot return them from functions. They decay to pointers, which then dangle.</p>
<p>C++11 introduced <a rel="noopener" href="https://en.cppreference.com/w/cpp/container/array" target="_blank"><code>std::array</code></a>, a statically-sized container, which solves this issue:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br></pre></td>
<td><pre><span></span><br><span>std::array&lt;int, 10&gt; create_arrray(){</span><br><span> <span>std</span>::<span>array</span>&lt;<span>int</span>, 10&gt; arr;</span><br><span> </span><br><span> <span>return</span> arr;</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<h3>
<a name="variadic-templates" href="#variadic-templates" title='Link to "Variadic templates"'>🔗</a>Variadic templates</h3>
<p>The missing piece of our puzzle: we can now have templates with any amount of template parameters, using parameter packs!</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br></pre></td>
<td><pre><span><span>template</span>&lt;<span>typename</span>... Args&gt;</span><br><span><span><span>void</span> <span>do_something</span><span>(Args&amp;&amp;... args)</span></span>{</span><br><span> </span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p>(If you’re confused by the <code>Args&amp;&amp;</code> nomenclature, check my previous article: <a title="Forwarding References and Perfect Forwarding in C++" href="https://joelfilho.com/blog/2020/forwarding_references/">Forwarding References and Perfect Forwarding in C++</a>)</p>
<h3>
<a name="compile-time-lut-in-c-11" href="#compile-time-lut-in-c-11" title='Link to "Compile-time LUT in C++11"'>🔗</a>Compile-time LUT in C++11</h3>
<p>We all know all the features we needed to create this, so let’s do it:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br><span>8</span><br><span>9</span><br><span>10</span><br><span>11</span><br><span>12</span><br><span>13</span><br><span>14</span><br><span>15</span><br><span>16</span><br><span>17</span><br><span>18</span><br><span>19</span><br><span>20</span><br><span>21</span><br><span>22</span><br><span>23</span><br><span>24</span><br><span>25</span><br><span>26</span><br><span>27</span><br><span>28</span><br><span>29</span><br><span>30</span><br><span>31</span><br><span>32</span><br><span>33</span><br></pre></td>
<td><pre><span></span><br><span></span><br><span></span><br><span></span><br><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length, <span>typename</span> T, <span>typename</span> Generator, <span>typename</span>... Values&gt;</span><br><span><span><span>constexpr</span> <span>auto</span> <span>lut_impl</span><span>(Generator&amp;&amp; f, Values... values)</span></span></span><br><span> -&gt; typename std::enable_if&lt;sizeof...(Values) == (Length - 1), std::array&lt;T, Length&gt;&gt;::type</span><br><span>{</span><br><span> <span>return</span> {values..., <span>std</span>::forward&lt;Generator&gt;(f)(<span>sizeof</span>...(Values))};</span><br><span>}</span><br><span></span><br><span></span><br><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length, <span>typename</span> T, <span>typename</span> Generator, <span>typename</span>... Values&gt;</span><br><span><span><span>constexpr</span> <span>auto</span> <span>lut_impl</span><span>(Generator&amp;&amp; f, Values... values)</span></span></span><br><span> -&gt; typename std::enable_if&lt;sizeof...(Values) &lt; (Length - 1), std::array&lt;T, Length&gt;&gt;::type</span><br><span>{</span><br><span> <span>return</span> lut_impl&lt;Length, T, Generator, Values..., <span>decltype</span>(f(<span>std</span>::<span>size_t</span>{<span>0</span>}))&gt;</span><br><span> (<span>std</span>::forward&lt;Generator&gt;(f),</span><br><span> values...,</span><br><span> f(<span>sizeof</span>...(Values)));</span><br><span>}</span><br><span></span><br><span></span><br><span></span><br><span></span><br><span></span><br><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length, <span>typename</span> Generator&gt;</span><br><span>constexpr auto lut(Generator&amp;&amp; f) -&gt; std::array&lt;decltype(f(std::size_t{0})), Length&gt; {</span><br><span> </span><br><span> <span>return</span> lut_impl&lt;Length, </span><br><span> <span>decltype</span>(f(<span>std</span>::<span>size_t</span>{<span>0</span>})) </span><br><span> &gt;(<span>std</span>::forward&lt;Generator&gt;(f));</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p>If you found this unreadable and extremely confusing, remember to thank your library implementers.
<strong>They do this kind of work so you don’t have to!</strong> All you have to do is use the library, which is definitely easier.</p>
<p><strong>Explanation</strong>:</p>
<ul>
<li>As we can only have a return type on C++11, we cannot just create the array in our function. So, we created a recursive helper function, to hold all our values until we construct the array.</li>
<li>We need to deduce the return type in each function. As we know it’s an array and its length, we can return <code>std::array&lt;decltype(f(std::size_t{0})), Length&gt;</code>.</li>
</ul>
<p><strong>Usage</strong>:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br></pre></td>
<td><pre><span></span><br><span><span><span>unsigned</span> <span>precomputed_factorial</span><span>(<span>unsigned</span> n)</span></span>{</span><br><span> <span>constexpr</span> <span>auto</span> lut = lut&lt;<span>10</span>&gt;(factorial);</span><br><span> <span>return</span> lut[n];</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p><a rel="noopener" href="https://godbolt.org/z/MGzW33" target="_blank">Try it out on Compiler Explorer</a>!</p>
<p><strong>Limitations</strong>:</p>
<ul>
<li>With <code>std::array</code>, we cannot legally access the data at compile time. So, lookup is only done at runtime.</li>
</ul>

<h2>
<a name="c-14-constexpr-rises" href="#c-14-constexpr-rises" title='Link to "C++14: constexpr rises"'>🔗</a>C++14: <code>constexpr</code> rises</h2>
<p>After a decade working on C++11, the committee decided taking a different approach: <strong>standards will be released in a 3-year cycle</strong>. This way, we don’t wait for the completion of some features, while many are complete and ready to be used.</p>
<p>C++14 was the first standard released this way, and mostly consisted of C++11 bug fixes.</p>
<h3>
<a name="down-with-the-annoying-constexpr-limitation" href="#down-with-the-annoying-constexpr-limitation" title='Link to "Down with the annoying constexpr limitation!"'>🔗</a>Down with the annoying <code>constexpr</code> limitation!</h3>
<p><code>constexpr</code> was a success. What the committee didn’t expect was how successful it would become.</p>
<p>The initial thought was “let’s replace macros with one-liners”. What actually happened? Compile-time programming became a staple of the language, a symbol of the zero-overhead principle.</p>
<p>Now, we can have multiple lines in a single <code>constexpr</code> function!</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br></pre></td>
<td><pre><span><span><span>constexpr</span> <span>unsigned</span> <span>factorial</span><span>(<span>unsigned</span> n)</span></span>{</span><br><span> <span>unsigned</span> result = <span>1</span>;</span><br><span> <span>for</span>(<span>unsigned</span> i = <span>2</span>; i &lt;= n; i++){</span><br><span> result *= i;</span><br><span> }</span><br><span> <span>return</span> result;</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p>Now, we can just build an <code>std::array</code> and… <strong>Oh, no!</strong> They forgot to make <code>std::array</code>‘s operators <code>constexpr</code>!</p>
<p>We could implement our own <code>constexpr notstd::array</code>, but, as the initial disclaimer, we’ll only use standard features. So, let’s see another addition to C++14.</p>
<h2>
<a name="std-integer-sequence-and-std-index-sequence" href="#std-integer-sequence-and-std-index-sequence" title='Link to "std::integer_sequence and std::index_sequence"'>🔗</a><code>std::integer_sequence</code> and <code>std::index_sequence</code>
</h2>
<p><a rel="noopener" href="https://en.cppreference.com/w/cpp/utility/integer_sequence" target="_blank"><code>std::integer_sequence</code> and <code>std::index_sequence</code></a> are template utilities for creating sequences and accessing multiple indexes as pack expansions.</p>
<h2>
<a name="a-better-auto" href="#a-better-auto" title='Link to "A better auto"'>🔗</a>A better <code>auto</code>
</h2>
<p>Now, the compiler can deduce the return type of our functions. No more trailing <code>decltype</code>!</p>
<p>Using our C++11 example, we now can remove our ugly <code>decltype</code> expression:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br></pre></td>
<td><pre><span><span>using</span> other_type = ;</span><br><span></span><br><span><span>template</span>&lt;<span>typename</span> F&gt;</span><br><span><span><span>auto</span> <span>call</span><span>(F&amp;&amp; f)</span> </span>{</span><br><span> other_type t { };</span><br><span> <span>return</span> f(t);</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p><strong>Note</strong>: <code>auto</code> implies returning a value. In order to return whichever type <code>f</code> returns, including references, we must define our return type <code>decltype(auto)</code>.</p>
<h3>
<a name="compile-time-lut-in-c-14" href="#compile-time-lut-in-c-14" title='Link to "Compile-time LUT in C++14"'>🔗</a>Compile-time LUT in C++14</h3>
<p>At the end, we have slightly more readable code:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br><span>8</span><br><span>9</span><br><span>10</span><br><span>11</span><br><span>12</span><br></pre></td>
<td><pre><span></span><br><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length, <span>typename</span> Generator, <span>std</span>::<span>size_t</span>... Indexes&gt;</span><br><span><span><span>constexpr</span> <span>auto</span> <span>lut_impl</span><span>(Generator&amp;&amp; f, <span>std</span>::index_sequence&lt;Indexes...&gt;)</span> </span>{</span><br><span> <span>using</span> content_type = <span>decltype</span>(f(<span>std</span>::<span>size_t</span>{<span>0</span>}));</span><br><span> <span>return</span> <span>std</span>::<span>array</span>&lt;content_type, Length&gt; {{ f(Indexes)... }};</span><br><span>}</span><br><span></span><br><span></span><br><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length, <span>typename</span> Generator&gt;</span><br><span><span><span>constexpr</span> <span>auto</span> <span>lut</span><span>(Generator&amp;&amp; f)</span></span>{</span><br><span> <span>return</span> lut_impl&lt;Length&gt;(<span>std</span>::forward&lt;Generator&gt;(f), <span>std</span>::make_index_sequence&lt;Length&gt;{});</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p><a rel="noopener" href="https://godbolt.org/z/TGzG4K" target="_blank">Try it out on Compiler Explorer</a>!</p>
<p><strong>Details</strong>:</p>
<ul>
<li>It is possible to <a rel="noopener" href="https://github.com/JoelFilho/JTC/blob/master/include/jtc/templates/integer_sequence.hpp" target="_blank">implement <code>std::index_sequence</code> on C++11</a>. The <code>decltype</code> trailing return types are still required there, though.</li>
<li>Like on C++11, lookup can still only be done at runtime.</li>
</ul>
<h2>
<a name="variable-templates" href="#variable-templates" title='Link to "Variable templates"'>🔗</a>Variable templates</h2>
<p>C++14 introduced <a rel="noopener" href="https://en.cppreference.com/w/cpp/language/variable_template" target="_blank">variable templates</a>. We now can declare our LUTs as any size, and use them as normal variables (<a rel="noopener" href="https://godbolt.org/z/eGqb84" target="_blank">See in Compiler Explorer</a>):</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br></pre></td>
<td><pre><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length&gt;</span><br><span><span>constexpr</span> <span>auto</span> factorial_lut = lut&lt;Length&gt;(factorial);</span><br><span></span><br><span></span><br><span><span><span>unsigned</span> <span>precomputed_factorial</span><span>(<span>unsigned</span> n)</span></span>{</span><br><span> <span>return</span> factorial_lut&lt;<span>10</span>&gt;[n];</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p><strong>Note</strong>: Only after C++17, we can declare our variable as <code>inline</code>. We may have some problems with linkage, as we need to take the address of a global variable inside our LUT. Multiple <code>static</code> variables may increase binary size, and not using static linkage may result in duplicate symbols.</p>

<h2>
<a name="c-17-constexpr-shines" href="#c-17-constexpr-shines" title='Link to "C++17: constexpr shines"'>🔗</a>C++17: <code>constexpr</code> shines</h2>
<h3>
<a name="a-fully-constexpr-std-array" href="#a-fully-constexpr-std-array" title='Link to "A fully constexpr std::array"'>🔗</a>A fully <code>constexpr</code> <code>std::array</code>
</h3>
<p>The committee finally added <code>constexpr</code> to <code>std::array</code>‘s iterators and <code>operator[]</code>. We can now iteratively construct an <code>std::array</code>:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br></pre></td>
<td><pre><span><span><span>constexpr</span> <span>auto</span> <span>random_array</span><span>()</span></span>{</span><br><span> <span>std</span>::<span>array</span>&lt;<span>int</span>, 10&gt; arr;</span><br><span> <span>for</span>(<span>auto</span>&amp; itm : arr){</span><br><span> itm = random_func();</span><br><span> }</span><br><span> <span>return</span> arr;</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<h3>
<a name="compile-time-lut-in-c-17" href="#compile-time-lut-in-c-17" title='Link to "Compile-time LUT in C++17"'>🔗</a>Compile-time LUT in C++17</h3>
<p>We can now go down to a single function!</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br><span>8</span><br><span>9</span><br><span>10</span><br><span>11</span><br></pre></td>
<td><pre><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length, <span>typename</span> Generator&gt;</span><br><span><span><span>constexpr</span> <span>auto</span> <span>lut</span><span>(Generator&amp;&amp; f)</span></span>{</span><br><span> <span>using</span> content_type = <span>decltype</span>(f(<span>std</span>::<span>size_t</span>{<span>0</span>}));</span><br><span> <span>std</span>::<span>array</span>&lt;content_type, Length&gt; arr {};</span><br><span></span><br><span> <span>for</span>(<span>std</span>::<span>size_t</span> i = <span>0</span>; i &lt; Length; i++){</span><br><span> arr[i] = f(i);</span><br><span> }</span><br><span></span><br><span> <span>return</span> arr;</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p><a rel="noopener" href="https://godbolt.org/z/T14MrK" target="_blank">Try it out on Compiler Explorer</a>!</p>
<p><strong>Note</strong>: This considers the result of <code>f</code> is default and copy constructible. Otherwise, we need to use the <code>index_sequence</code> version.</p>
<h3>
<a name="constexpr-lambdas" href="#constexpr-lambdas" title='Link to "constexpr lambdas"'>🔗</a><code>constexpr</code> lambdas</h3>
<p>Lambda expressions in C++17 now can be declared <code>constexpr</code>, to indicate its <code>operator()</code> can be called at compile time. It can also just infer <code>constexpr</code> from it. </p>
<p>With lambdas and <code>inline</code> variable templates, we can now declare our LUT in a single statement:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br><span>8</span><br></pre></td>
<td><pre><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length&gt;</span><br><span><span>inline</span> <span>constexpr</span> <span>auto</span> factorial_lut = lut&lt;Length&gt;([](<span>std</span>::<span>size_t</span> n){</span><br><span> <span>unsigned</span> result = <span>1</span>;</span><br><span> <span>for</span>(<span>unsigned</span> i = <span>2</span>; i &lt;= n; i++){</span><br><span> result *= i;</span><br><span> }</span><br><span> <span>return</span> result;</span><br><span>});</span><br></pre></td>
</tr></tbody></table></figure>

<h2>
<a name="c-20-constexpr-unleashed" href="#c-20-constexpr-unleashed" title='Link to "C++20: constexpr unleashed"'>🔗</a>C++20: <code>constexpr</code> unleashed</h2>
<p>C++20 brings the biggest upgrades in C++ since C++11. However, we’ll just use a couple of features.</p>
<h3>
<a name="constexpr-algorithms-and-ranges" href="#constexpr-algorithms-and-ranges" title='Link to "constexpr algorithms and ranges"'>🔗</a><code>constexpr</code> algorithms and ranges</h3>
<p>In his classic <a rel="noopener" href="https://channel9.msdn.com/Events/GoingNative/2013/Cpp-Seasoning" target="_blank">C++ Seasoning talk</a>, Sean Parent introduced us to the concept of “no raw loops”.</p>
<p>So, as good listeners, let’s use the fact that C++20 introduces <code>constexpr</code> to the functions on the <code>&lt;algorithm&gt;</code> header!</p>
<p>C++20 also introduced <a rel="noopener" href="https://en.cppreference.com/w/cpp/ranges" target="_blank">the ranges library</a>. It’s enormous, and changes the way we think about algorithms in C++. We’ll use them, rather than the usual algorithm functions.</p>
<h3>
<a name="compile-time-lut-in-c-20-feat-ranges" href="#compile-time-lut-in-c-20-feat-ranges" title='Link to "Compile-time LUT in C++20, feat. ranges"'>🔗</a>Compile-time LUT in C++20, feat. ranges</h3>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br><span>8</span><br><span>9</span><br><span>10</span><br><span>11</span><br><span>12</span><br></pre></td>
<td><pre><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length, <span>typename</span> Generator&gt;</span><br><span><span><span>constexpr</span> <span>auto</span> <span>lut</span><span>(Generator&amp;&amp; f)</span></span>{</span><br><span> <span>using</span> content_type = <span>decltype</span>(f(<span>std</span>::<span>size_t</span>{<span>0</span>}));</span><br><span> <span>std</span>::<span>array</span>&lt;content_type, Length&gt; arr;</span><br><span></span><br><span> <span>using</span> <span>namespace</span> <span>std</span>::ranges;</span><br><span> <span>auto</span> content = views::iota(<span>std</span>::<span>size_t</span>{<span>0</span>}, Length) </span><br><span> | views::transform(<span>std</span>::forward&lt;Generator&gt;(f)); </span><br><span> copy(content, arr.begin());</span><br><span></span><br><span> <span>return</span> arr;</span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p><a rel="noopener" href="https://godbolt.org/z/qKhcc7" target="_blank">Try it out on Compiler Explorer</a>!</p>
<p>Okay, that isn’t easier to read than C++17’s. (No, using <code>std::ranges::transform</code> won’t make it more readable.)</p>
<p>But that’s the beauty: C++ is a backwards-compatible language, that one still works!
And the language is highly complex, so this minimal example won’t be fair to the feature. Listen to Sean Parent, use algorithms.</p>

<h2>
<a name="what-about-the-future" href="#what-about-the-future" title='Link to "What about the future?"'>🔗</a>What about the future?</h2>
<p>We saw C++20 added ranges. Sadly, they didn’t introduce adapters to generate our data types. Hopefully, in the future, we can do something like this:</p>
<figure><table><tbody><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br></pre></td>
<td><pre><span><span>template</span>&lt;<span>std</span>::<span>size_t</span> Length, <span>typename</span> Generator&gt;</span><br><span><span><span>constexpr</span> <span>auto</span> <span>lut</span><span>(Generator&amp;&amp; f)</span></span>{</span><br><span> <span>using</span> <span>namespace</span> <span>std</span>::ranges;</span><br><span> <span>return</span> views::iota(<span>0u</span>z, Length) |&gt; views::transform(f) |&gt; to_array();</span><br><span> </span><br><span>}</span><br></pre></td>
</tr></tbody></table></figure>
<p><strong>Notes</strong>: </p>
<ul>
<li>
<a rel="noopener" href="https://wg21.link/P0330" target="_blank">P0330</a>: <code>uz</code> is the newly approved suffix for <code>size_t</code> literals!</li>
<li>
<a rel="noopener" href="https://wg21.link/P2011" target="_blank">P2011</a>: The “pizza operator” <code>|&gt;</code> is a solution proposed for an unwanted overhead of ranges</li>
<li>
<a rel="noopener" href="https://wg21.link/p1206" target="_blank">P1206</a>: “<code>ranges::to</code>: A function to convert any range to a container”. It says “Future work is needed to allow constructing <code>std::array</code>“</li>
</ul>

<h2>
<a name="conclusion" href="#conclusion" title='Link to "Conclusion"'>🔗</a>Conclusion</h2>
<p>Only by looking back, we can see how far we’ve come. With <code>constexpr</code>, it’s no different.</p>
<p>For our application, we found a balance. We can now generate lookup tables beautifully in our language, without depending on other languages, and at no runtime cost!</p>
<p>At the end, I still don’t like that we cannot use the C++17/C++20 versions for types that are not default constructible. We still need to use the C++14 version for those types, maybe only adding CTAD.</p>
<p>Perhaps a <code>to_array</code> adapter will work someday. I tried implementing it with my small knowledge of ranges, but I believe it may need <a rel="noopener" href="https://wg21.link/P1045" target="_blank">P1045</a>, so we can deduce the array length from a function input. Until then, we’ll keep using what we can.</p>
<p>If you came so far, thanks for reading! I have DMs open <a rel="noopener" href="https://twitter.com/_JoelFilho" target="_blank">on twitter</a>, and <a rel="noopener" href="https://www.reddit.com/user/JoelFilho/" target="_blank">on Reddit</a>. Any questions or comments, just let me know!</p>
<p>The larger code snippets from this post are released under the Boost License, on my <a rel="noopener" href="https://github.com/JoelFilho/blog_samples/tree/main/compile_time_lookup_tables" target="_blank">snippets repository</a>.</p>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via Joel Filho https://joelfilho.com<br>
<br>
December 1, 2023 at 02:04PM
